lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp

```diff
@@ -126,6 +126,104 @@ Value createLLVMIntegerConstant(OpBuilder &builder, Location loc, short width,
+      LLVM::LLVMPointerType::get(IntegerType::get(context, 8))};
+  auto funcType = LLVM::LLVMFunctionType::get(retType, argsType);
+
+  ConversionPatternRewriter::InsertionGuard guard(builder);
+  builder.setInsertionPointToStart(moduleOp.getBody());
+
+  return builder.create<LLVMFuncOp>(UnknownLoc::get(context), funcName,
+                                    funcType);
+}
+
+void llPrintf(StringRef msg, ValueRange args, OpBuilder &builder) {
+
+  static const char formatStringPrefix[] = "printfFormat_";
+  assert(!msg.empty() && "printf with empty string not support");
+
+  auto *context = builder.getContext();
+  auto moduleOp = builder.getBlock()->getParent()->getParentOfType<ModuleOp>();
+  auto funcOp = getVprintfDeclaration(builder);
+
+  Type int8Ty = builder.getIntegerType(8);
+  Type int8Ptr = LLVM::LLVMPointerType::get(int8Ty);
+  Type int32Ty = builder.getIntegerType(32);
+  Value one = builder.create<LLVM::ConstantOp>(
+      UnknownLoc::get(context), int32Ty, builder.getI32IntegerAttr(1));
+  Value zero = builder.create<LLVM::ConstantOp>(
+      UnknownLoc::get(context), int32Ty, builder.getI32IntegerAttr(0));
+
+  unsigned stringNumber = 0;
+  SmallString<16> stringConstName;
+  do {
+    stringConstName.clear();
+    (formatStringPrefix + Twine(stringNumber++)).toStringRef(stringConstName);
+  } while (moduleOp.lookupSymbol(stringConstName));
+
+  llvm::SmallString<20> formatString(msg);
+  formatString.push_back('\0');
+  size_t formatStringSize = formatString.size_in_bytes();
+  auto globalType = LLVM::LLVMArrayType::get(int8Ty, formatStringSize);
+
+  LLVM::GlobalOp global;
+  {
+    ConversionPatternRewriter::InsertionGuard guard(builder);
+    builder.setInsertionPointToStart(moduleOp.getBody());
+    global = builder.create<LLVM::GlobalOp>(
+        UnknownLoc::get(context), globalType,
+        /*isConstant=*/true, LLVM::Linkage::Internal, stringConstName,
+        builder.getStringAttr(formatString));
+  }
+
+  Value globalPtr =
+      builder.create<LLVM::AddressOfOp>(UnknownLoc::get(context), global);
+  Value stringStart =
+      builder.create<LLVM::GEPOp>(UnknownLoc::get(context), int8Ptr, globalPtr,
+                                  mlir::ValueRange({zero, zero}));
+
+  Value bufferPtr =
+      builder.create<LLVM::NullOp>(UnknownLoc::get(context), int8Ptr);
+  if (args.size() >= 1) {
+    SmallVector<Type> argTypes;
+    for (auto arg : args) {
+      argTypes.push_back(arg.getType());
+    }
+    Type structTy = LLVM::LLVMStructType::getLiteral(context, argTypes);
+    auto allocated = builder.create<LLVM::AllocaOp>(
+        UnknownLoc::get(context), LLVM::LLVMPointerType::get(structTy), one,
+        /*alignment=*/0);
+
+    for (const auto &entry : llvm::enumerate(args)) {
+      auto index = builder.create<LLVM::ConstantOp>(
+          UnknownLoc::get(context), int32Ty,
+          builder.getI32IntegerAttr(entry.index()));
+      auto fieldPtr = builder.create<LLVM::GEPOp>(
```

Comment:
Ditto, there is a macro called `gep` which is short for `builder.create<LLVM::GEPOp>`

Reply:
Good suggestion, thanks.